### PR TITLE
MgrMonitor: consider the build config when defining the always-on modules

### PR DESCRIPTION
**Pull request by leonid-s-usov**
Following a [discussion on slack](https://ceph-storage.slack.com/archives/C1HFZTW81/p1702900067810509)

> When ceph is built `WITH_RBD=OFF` shouldn’t we also disable the `rbd_support` mgr module? I suspect that this causes the health warning for me:
> ```
> "Module 'rbd_support' has failed dependency: module 'rbd' has no attribute 'Image'"
> ```

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/mon/MgrMonitor.cc
Comment (idryomov) [inline src/mon/MgrMonitor.cc:76]: What is the reason for introducing `std::call_once()` complexity?  Wouldn't a regular function-local static work and be more efficient?  Something like:
```
  static std::set<std::string> always_on_modules = {
    "crash",
    "status",
    "progress",
    "balancer",
    "devicehealth",
    "orchestrator",
#ifdef WITH_RBD
    "rbd_support",
#endif
#ifdef WITH_CEPHFS
    "volumes",
#endif
    "pg_autoscaler",
    "telemetry",
  };
  static std::map<uint32_t, std::set<std::string>> always_on_modules_map = {
    {CEPH_RELEASE_OCTOPUS, always_on_modules},
    {CEPH_RELEASE_PACIFIC, always_on_modules},
    {CEPH_RELEASE_QUINCY, always_on_modules},
    {CEPH_RELEASE_REEF, always_on_modules},
  };
  return always_on_modules_map;
```

File: src/mon/MgrMonitor.cc
Comment (idryomov) [inline src/mon/MgrMonitor.cc:89]: Nit: don't indent `ifdef` itself and don't indent _based on_ `ifdef`.  This would be better formatted as:

```suggestion
#ifdef WITH_CEPHFS
    octopus_modules.insert("volumes");
#endif
```
One (or their editor/IDE) should be able to hide preprocessor directives (picking one of the branches) and not have to reindent everything.

File: src/mon/MgrMonitor.cc
Comment (idryomov) [inline src/mon/MgrMonitor.cc:97]: Nit: naming a common set of modules after a particular release is a bit confusing -- one immediately wonders whether ` always_on_modules[CEPH_RELEASE_PACIFIC] = octopus_modules` is a cut-and-paste typo.  I would suggest `common_modules` or `always_on_modules` for the set and `always_on_modules_map` for the "release" map.

Comment (leonid-s-usov) [reply src/mon/MgrMonitor.cc:97]: I thought about this a little and chose this convention to emphasize the fact that the set is the same for all releases. Now that you mention it I agree that it may appear confusing more than telling.

TBH, I'm a little disappointed by the redundancy of this structure. I'd only introduce a new set when it is really different, and look up by the `lower_bound` of the current release.

File: src/mon/MgrMonitor.cc
Comment (leonid-s-usov) [reply src/mon/MgrMonitor.cc:76]: Well, this approach is indeed simpler. However, I'd prefer it with my current naming convention of `octopus_modules`, because if the next release expands on the list it won't be able to reuse the same set and would need to come up with another name or encode it in place.

Comment (leonid-s-usov) [reply src/mon/MgrMonitor.cc:76]: anyway, I like this option, will pick the suggestion

File: src/mon/MgrMonitor.cc
Comment (idryomov) [inline src/mon/MgrMonitor.cc:73]: I'm not sure I see a reason for appending `release_` to the name -- it no longer matches the name of mgrmap field.  But naming here is up to @ceph/core.

**Review by idryomov — APPROVED** [review]
Ack from the RBD perspective

File: src/mon/MgrMonitor.cc
Comment (leonid-s-usov) [reply src/mon/MgrMonitor.cc:73]: If `always_on_modules` is a list/set, then this is an association between a release and the set. But `always_on_modules_map` is ambiguous as a generic name IMO since it doesn't clarify the discriminator. Anyway, I could pick any other name following the code maintainers' preference, and I was kind of forced to pick some since `always_on_modules` was the name of the static variable in this function before my last edit - which isn't the case now, so I could revert. Will do it if needed.

**Comment by leonid-s-usov** [comment]
jenkins test make check

**Comment by leonid-s-usov** [comment]
jenkins test make check arm64

**Comment by leonid-s-usov** [comment]
@idryomov do you think this can be merged? Should we run a test suite on this?

**Comment by leonid-s-usov** [comment]
jenkins test api

**Comment by idryomov** [comment]
> @idryomov do you think this can be merged? Should we run a test suite on this?

I think this should be reviewed by @ceph/core and go through the rados suite.

**Comment by leonid-s-usov** [comment]
Who could review this from the core team? Also, how can this be scheduled for the rados suite?

**Comment by idryomov** [comment]
> Who could review this from the core team? Also, how can this be scheduled for the rados suite?

Perhaps @ljflores or @rzarzynski?

File: src/mon/MgrMonitor.cc
Comment (rzarzynski) [inline src/mon/MgrMonitor.cc:81]: ACK.

File: src/mon/MgrMonitor.cc
Comment (rzarzynski) [inline src/mon/MgrMonitor.cc:95]: ACK. I'm assuming before the changes there was no differences in module list between the versions.

**Review by rzarzynski — APPROVED** [review]
Testing this isn't going to be easy as we need an image with non-standard parameters. Two options that came to my mind are:

1. Check whether jenkins allows to add non-standards args (like `-DWITH_RBD=off`) to cmake invocation.
2. Add a patch-just-for-testing that switches in the default value of the params in `CMakeList.txt`.

File: src/mon/MgrMonitor.cc
Comment (leonid-s-usov) [reply src/mon/MgrMonitor.cc:95]: Correct. This way it's obvious and concise, previously it was a copy-paste of the same list.

**Comment by leonid-s-usov** [comment]
> Testing this isn't going to be easy as we need an image with non-standard parameters. Two options that came to my mind are:
> 
> 1. Check whether jenkins allows to add non-standards args (like `-DWITH_RBD=off`) to cmake invocation.
> 2. Add a patch-just-for-testing that switches in the default value of the params in `CMakeList.txt`.

I can create an image following option 2. What suite should be run so that the absence of rbd wouldn't be an issue for them?

**Comment by idryomov** [comment]
> Testing this isn't going to be easy as we need an image with non-standard parameters.

I don't think we need to test with non-standard build parameters in teuthology.  This is mostly for development purposes, so to smoke test `-DWITH_RBD=OFF` and/or `-DWITH_CEPHFS=OFF`, a local build should suffice.  I assume Leonid has already done that since being able to do that in a development setup is the entire purpose of this PR.

**Comment by leonid-s-usov** [comment]
> > Testing this isn't going to be easy as we need an image with non-standard parameters.
> 
> I don't think we need to test with non-standard build parameters in teuthology. This is mostly for development purposes, so to smoke test `-DWITH_RBD=OFF` and/or `-DWITH_CEPHFS=OFF`, a local build should suffice. I assume Leonid has already done that since being able to do that in a development setup is the entire purpose of this PR.

I will try with both off, until now I only excluded the RBD

**Comment by leonid-s-usov** [comment]
built with the settings:

```
./do_cmake.sh -DCMAKE_BUILD_TYPE=RelWithDebInfo -DWITH_MGR_DASHBOARD_FRONTEND=OFF -DWITH_RBD=OFF -DWITH_CEPHFS=OFF -DWITH_KRBD=OFF
```

started with `MON=1 MDS=0 OSD=1 ../src/vstart.sh -n -l`

```
[leonidus@vossi04 build]$ ceph -s
2024-04-04T23:56:47.325+0000 7f68827c2700 -1 WARNING: all dangerous and experimental features are enabled.
2024-04-04T23:56:47.328+0000 7f68827c2700 -1 WARNING: all dangerous and experimental features are enabled.
  cluster:
    id:     82485ca7-09ce-4eee-8675-17f113b76371
    health: HEALTH_WARN
            1 pool(s) have no replicas configured

  services:
    mon: 1 daemons, quorum a (age 20m)
    mgr: x(active, since 19m)
    osd: 1 osds: 1 up (since 19m), 1 in (since 19m)

  data:
    pools:   1 pools, 1 pgs
    objects: 2 objects, 577 KiB
    usage:   1.0 GiB used, 100 GiB / 101 GiB avail
    pgs:     1 active+clean
```

Let me know if it's good to merge 🙏🏻

**Comment by idryomov** [comment]
Singe this is mostly a mechanical change that has been pending for a long time, I'm merging based on Radek's approval and Leonid's local testing.

**Pull request merged**
